### PR TITLE
Discover Plugins from all active contexts and Support Plugin Name conflicts across different Targets

### PR DESCRIPTION
**Pull request by anujc25**
### What this PR does / why we need it

* Discover Plugins from all active contexts:
    - Before this change, Tanzu CLI only discovers plugins based on the `current` (currentServer) field. But with targets implementation, we can have one active context per target. Meaning, we need to discover plugins from multiple active contexts.
    - Changes are made in `pluginmanager` APIs to not take `serverName` as parameter but fetch the active contexts as part of internal `DiscoverServerPlugins` function and provide the result based on the active contexts instead of just single `current` (currentServer) field.

* Support Plugin Name conflicts across different Targets
    - Before this change, Tanzu CLI uniquely identifies plugins based on just the `PluginName` irrespective of whether it is context-scoped plugin, standalone plugin or belongs to a different target group. Meaning, only one plugin of a name can be installed at a time. However, there is expectation that there can be different plugins of the same name (regardless of how they are discovered), each associated with a different target level. 
    - This change, addresses this issue by identifying uniqueness based on `PluginName` and `Target` of the plugin. This means, plugin names must be Unique per Target Level.

Below are the high-level summary of the changes:
- Add `Target` field in CLIPlugin CR for determining Target of the standalone plugins
- Update `Catalog` to use `PluginName_Target` as a key to store plugin details in the map
- Update `plugin-manager` lifecycle APIs to not accept `ServerName` as input but find the current server information as part of `DiscoverServerPlugins` API. If the `context-target` feature-flag is enabled, `DiscoverServerPlugins` API will return plugin discovered through all active contexts.
- Add `target` flag for plugin lifecycle commands to uniquely identify the plugins incase of name conflicts. 
- Add `target` information for standalone plugins CR defined under `standalone-plugins` package
- Shortened format handling for `k8s` targeted plugins when there is name conflict between `<none>` targeted plugin and `k8s` targeted plugin.
  - Special handling of `k8s` target plugins is needed here as both the plugins can be treated as same plugins
- Update implementation of builder plugin to accept target information with the plugin name and to generate CLIPlugin resource containing `target` information
- Add/Update some of the docs

TODO:
- Update `Context-Type` in ClientConfig API to `Target`. (This will be done as part of follow-up PR)

### Which issue(s) this PR fixes

Related to #3714

### Describe testing done for PR

**Plugin Install Tests**

List the plugins before installing any
```
~/.c/tanzu-plugins $ tanzu plugin list
Standalone Plugins
  NAME                DESCRIPTION                                                        TARGET      DISCOVERY      VERSION      STATUS         
  login               Login to the platform                                                          default-local  v0.28.0-dev  not installed  
  pinniped-auth       Pinniped authentication operations (usually not directly invoked)              default-local  v0.28.0-dev  not installed  
  management-cluster  Kubernetes management cluster operations                           kubernetes  default-local  v0.28.0-dev  not installed  
  package             Tanzu package management                                           kubernetes  default-local  v0.28.0-dev  not installed  
  secret              Tanzu secret management                                            kubernetes  default-local  v0.28.0-dev  not installed  
  telemetry           configure cluster-wide settings for vmware tanzu telemetry         kubernetes  default-local  v0.28.0-dev  not installed  

Plugins from Context:  kind-test-cluster
  NAME                DESCRIPTION                           TARGET      VERSION      STATUS         
  cluster             Kubernetes cluster operations         kubernetes  v0.28.0-dev  not installed  
  feature             Operate on features and featuregates  kubernetes  v0.28.0-dev  not installed  
  kubernetes-release  Kubernetes release operations         kubernetes  v0.28.0-dev  not installed  

Plugins from Context:  tmc-unstable
  NAME                DESCRIPTION                                                     TARGET           VERSION  STATUS         
  account             Account for tmc resources                                       mission-control  v0.0.1   not installed  
  apply               Create/Update a resource with a resource file                   mission-control  v0.0.1   not installed  
  audit               Run an audit request on an org                                  mission-control  v0.0.1   not installed  
  cluster                                                                             mission-control  v0.0.1   not installed  
  clustergroup        A group of Kubernetes clusters                                  mission-control  v0.0.1   not installed  
  data-protection     Data protection for tmc resources                               mission-control  v0.0.1   not installed  
  ekscluster                                                                          mission-control  v0.0.1   not installed  
  events              Events for any meaningful user activity or system state change  mission-control  v0.0.1   not installed  
  iam                 IAM Policies for tmc resources                                  mission-control  v0.0.1   not installed  
  inspection          Inspection for tmc resources                                    mission-control  v0.0.1   not installed  
  integration         Get available integrations and their information from registry  mission-control  v0.0.1   not installed  
  management-cluster  A TMC registered Management cluster                             mission-control  v0.0.1   not installed  
  policy              Policy for tmc resources                                        mission-control  v0.0.1   not installed  
  workspace           A group of Kubernetes namespaces                                mission-control  v0.0.1   not installed  
```
Different Plugin Name conflict scenarios 
```
~/.c/tanzu-plugins $ tanzu plugin install login
ℹ  Installing plugin 'login:v0.28.0-dev'
✔  successfully installed 'login' plugin

~/.c/tanzu-plugins $ tanzu plugin install cluster
✖  unable to uniquely identify plugin 'cluster'. Please specify correct Target(kubernetes[k8s]/mission-control[tmc]) of the plugin with `--target` flag 

~/.c/tanzu-plugins [1]$ tanzu plugin install cluster --target k8s
ℹ  Installing plugin 'cluster:v0.28.0-dev' with target 'kubernetes'
✔  successfully installed 'cluster' plugin
```

Check `tanzu --help` to see the `cluster` command is available under root tanzu cli command as well as under `kubernetes` target with `tanzu kubernetes --help`

```
~/.c/tanzu-plugins $ tanzu 
Tanzu CLI

Usage:
  tanzu [command]

Available command groups:

  Run
    cluster                 Kubernetes cluster operations 

  System
    completion              Output shell completion code 
    config                  Configuration for the CLI 
    context                 Configure and manage contexts for the Tanzu CLI 
    init                    Initialize the CLI 
    plugin                  Manage CLI plugins 
    update                  Update the CLI 
    version                 Version information 

  Target
    kubernetes              Tanzu CLI plugins that target a Kubernetes cluster 
    mission-control         Tanzu CLI plugins that target a Tanzu Mission Control endpoint 
``` 
```
~/.c/tanzu-plugins $ tanzu cluster info
{"name":"cluster","description":"Kubernetes cluster operations","version":"v0.28.0-dev","buildSHA":"522e7a192-dirty","digest":"","group":"Run","docURL":"","completionType":0,"aliases":["cl","clusters"],"installationPath":"","discovery":"","scope":"","status":"","discoveredRecommendedVersion":"","target":"","defaultFeatureFlags":{"features.cluster.custom-nameservers":false,"features.cluster.dual-stack-ipv4-primary":false,"features.cluster.dual-stack-ipv6-primary":false},"pluginRuntimeVersion":"v0.28.0-dev"}
```
```
~/.c/tanzu-plugins $ tanzu k8s --help
Tanzu CLI plugins that target a Kubernetes cluster

Usage:
  tanzu kubernetes [command]

Available command groups:

  Run
    cluster                 Kubernetes cluster operations 

```
```
~/.c/tanzu-plugins $ tanzu k8s cluster info
{"name":"cluster","description":"Kubernetes cluster operations","version":"v0.28.0-dev","buildSHA":"522e7a192-dirty","digest":"","group":"Run","docURL":"","completionType":0,"aliases":["cl","clusters"],"installationPath":"","discovery":"","scope":"","status":"","discoveredRecommendedVersion":"","target":"","defaultFeatureFlags":{"features.cluster.custom-nameservers":false,"features.cluster.dual-stack-ipv4-primary":false,"features.cluster.dual-stack-ipv6-primary":false},"pluginRuntimeVersion":"v0.28.0-dev"}
```

Continue installing plugin from tmc target:
```
~/.c/tanzu-plugins $ tanzu plugin install cluster --target tmc
ℹ  Installing plugin 'cluster:v0.0.1' with target 'mission-control'
✔  successfully installed 'cluster' plugin
```

Verify the help to see the `cluster` command is available under `tmc` target
```
~/.c/tanzu-plugins $ tanzu tmc --help
Tanzu CLI plugins that target a Tanzu Mission Control endpoint

Usage:
  tanzu mission-control [command]

Available command groups:

  Manage
    cluster                 A TMC managed Kubernetes cluster 
```
```
~/.c/tanzu-plugins $ tanzu tmc cluster info
{"name":"cluster","description":"A TMC managed Kubernetes cluster","version":"v0.0.1","buildSHA":"9d6682aa","digest":"","group":"Manage","docURL":"","completionType":0,"installationPath":"","discovery":"","scope":"","status":"","discoveredRecommendedVersion":""}
```

List the plugins after installing few plugins.
```
~/.c/tanzu-plugins $ tanzu plugin list
Standalone Plugins
  NAME                DESCRIPTION                                                        TARGET      DISCOVERY      VERSION      STATUS         
  login               Login to the platform                                                          default-local  v0.28.0-dev  installed      
  pinniped-auth       Pinniped authentication operations (usually not directly invoked)              default-local  v0.28.0-dev  not installed  
  management-cluster  Kubernetes management cluster operations                           kubernetes  default-local  v0.28.0-dev  not installed  
  package             Tanzu package management                                           kubernetes  default-local  v0.28.0-dev  not installed  
  secret              Tanzu secret management                                            kubernetes  default-local  v0.28.0-dev  not installed  
  telemetry           configure cluster-wide settings for vmware tanzu telemetry         kubernetes  default-local  v0.28.0-dev  not installed  

Plugins from Context:  kind-test-cluster
  NAME                DESCRIPTION                           TARGET      VERSION      STATUS         
  cluster             Kubernetes cluster operations         kubernetes  v0.28.0-dev  installed      
  feature             Operate on features and featuregates  kubernetes  v0.28.0-dev  not installed  
  kubernetes-release  Kubernetes release operations         kubernetes  v0.28.0-dev  not installed  

Plugins from Context:  tmc-unstable
  NAME                DESCRIPTION                                                     TARGET           VERSION  STATUS         
  account             Account for tmc resources                                       mission-control  v0.0.1   not installed  
  apply               Create/Update a resource with a resource file                   mission-control  v0.0.1   not installed  
  audit               Run an audit request on an org                                  mission-control  v0.0.1   not installed  
  cluster                                                                             mission-control  v0.0.1   installed      
  clustergroup        A group of Kubernetes clusters                                  mission-control  v0.0.1   not installed  
  data-protection     Data protection for tmc resources                               mission-control  v0.0.1   not installed  
  ekscluster                                                                          mission-control  v0.0.1   not installed  
  events              Events for any meaningful user activity or system state change  mission-control  v0.0.1   not installed  
  iam                 IAM Policies for tmc resources                                  mission-control  v0.0.1   not installed  
  inspection          Inspection for tmc resources                                    mission-control  v0.0.1   not installed  
  integration         Get available integrations and their information from registry  mission-control  v0.0.1   not installed  
  management-cluster  A TMC registered Management cluster                             mission-control  v0.0.1   not installed  
  policy              Policy for tmc resources                                        mission-control  v0.0.1   not installed  
  workspace           A group of Kubernetes namespaces                                mission-control  v0.0.1   not installed  
```

- Plugin Upgrade Tests

Different Plugin Name conflict scenarios 
```
~/.c/tanzu-plugins $ tanzu plugin upgrade login
ℹ  Installing plugin 'login:v0.28.0-dev'
✔  successfully upgraded plugin 'login' to version 'v0.28.0-dev'

~/.c/tanzu-plugins $ tanzu plugin upgrade cluster
✖  unable to uniquely identify plugin 'cluster'. Please specify correct Target(kubernetes[k8s]/mission-control[tmc]) of the plugin with `--target` flag 

~/.c/tanzu-plugins [1]$ tanzu plugin upgrade cluster --target k8s
ℹ  Installing plugin 'cluster:v0.28.0-dev' with target 'kubernetes'
✔  successfully upgraded plugin 'cluster' to version 'v0.28.0-dev'

~/.c/tanzu-plugins $ tanzu plugin install cluster --target tmc
ℹ  Installing plugin 'cluster:v0.0.1' with target 'mission-control'
✔  successfully upgraded plugin 'cluster' to version 'v0.0.1'
```

- Plugin Describe Tests
```
~/.c/tanzu-plugins [1]$ tanzu plugin describe login
... <shows correct output>

~/.c/tanzu-plugins $ tanzu plugin describe cluster
✖  unable to uniquely identify plugin 'cluster'. Please specify correct Target(kubernetes[k8s]/mission-control[tmc]) of the plugin with `--target` flag 

~/.c/tanzu-plugins [1]$ tanzu plugin describe cluster --target k8s
... <shows correct output>

~/.c/tanzu-plugins [1]$ tanzu plugin describe cluster --target tmc
... <shows correct output>
```
- Plugin Delete Tests
```
~/.c/tanzu-plugins $ tanzu plugin delete login
Deleting Plugin 'login'. Are you sure? [y/N]: y
✔  successfully deleted plugin 'login

~/.c/tanzu-plugins $ tanzu plugin delete cluster
✖  unable to uniquely identify plugin 'cluster'. Please specify correct Target(kubernetes[k8s]/mission-control[tmc]) of the plugin with `--target` flag 

~/.c/tanzu-plugins [1]$ tanzu plugin delete cluster --target kubernetes
Deleting Plugin 'cluster'. Are you sure? [y/N]: y
✔  successfully deleted plugin 'cluster'

~/.c/tanzu-plugins $ tanzu plugin delete cluster --target tmc
Deleting Plugin 'cluster'. Are you sure? [y/N]: y
✔  successfully deleted plugin 'cluster'
```
- Context List
```
~/.c/tanzu-plugins $ tanzu context list 
Target:  k8s
  NAME               ISACTIVE  ENDPOINT  KUBECONFIGPATH             KUBECONTEXT        
  kind-test-cluster  true                /Users/anujc/.kube/config  kind-test-cluster  
Target:  tmc
  NAME          ISACTIVE  ENDPOINT                               
  tmc           false     unstable-dev.tmc.cloud.vmware.com:443  
  tmc-unstable  true      unstable.tmc-dev.cloud.vmware.com:443 
```

### Release note

```release-note
Support Plugin Name conflicts across different Targets
```

### CODE REVIEW

File: cli/core/docs/cli/cli-architecture.md
Comment (marckhouzam) [inline cli/core/docs/cli/cli-architecture.md:177]: The second sentence is hard to understand.  How about replacing the second sentence with:
"To respect backwards-compatibility, when no target command is specified, the `kubernetes` target is used automatically.

File: cli/core/docs/cli/cli-architecture.md
Comment (marckhouzam) [inline cli/core/docs/cli/cli-architecture.md:186]: This sentence needs an introduction.  How about changing things to:
"Using the `target` concept, we can differentiate between listing TKG workload clusters or TMC workload clusters.
To list TKG workload clusters using the TKG `cluster` plugin which is associated with the `kubernetes` target:"

File: cli/core/docs/cli/cli-architecture.md
Comment (marckhouzam) [inline cli/core/docs/cli/cli-architecture.md:196]: If you decide to take my suggestion above, the matching would be here:
"To list TMC workload clusters using the TMC `cluster` plugin which is associated with the `tmc` target:"

File: cli/core/docs/cli/cli-architecture.md
Comment (marckhouzam) [inline cli/core/docs/cli/cli-architecture.md:205]: "... hidden commands." -> "... hidden commands such as the `info` command as shown in the next example."

**Review by marckhouzam — COMMENTED** [review]
Thanks for the change @anujc25, I'm starting with reviewing the docs and posting comments immediately to go faster

File: cli/core/pkg/pluginmanager/manager.go
Comment (marckhouzam) [inline cli/core/pkg/pluginmanager/manager.go:226]: Is it ok that we no longer specify an empty server here?  IIUC before we used to request only the standalone plugins since the server passed was empty, but now what will happen since we can't specify a server name anymore?

**Review by marckhouzam — COMMENTED** [review]
Continuing the review.  Things are coming together 😄 
Here are some comments/questions for things I was wondering.
Will continue to review.

**Comment by marckhouzam** [comment]
I noticed that `tanzu context create ...` seems to trigger plugins to be installed for all active contexts, meaning the new one I created, but also for the other target.  I wonder if this is right?  

Say I use a TMC target and remove some plugins I don't want to use (`tanzu plugin delete`); then I create a context to point to a TKG cluster but then all my TMC plugins get re-installed.

File: cli/core/pkg/command/discovery_source.go
Comment (anujc25) [reply cli/core/pkg/command/discovery_source.go:75]: Good catch. We should be displaying this based on all active contexts when the feature-flag is enabled. I will update this.

File: cli/core/pkg/command/root.go
Comment (anujc25) [reply cli/core/pkg/command/root.go:114]: Yeah. This should be pointing to `common.PluginScopeContext`. I will update this.

File: cli/core/pkg/command/root.go
Comment (anujc25) [reply cli/core/pkg/command/root.go:117]: This is great point. I didn't think about the core commands. These are good suggestions. I will think a bit more and get back on this.

File: cli/core/pkg/command/root.go
Comment (anujc25) [reply cli/core/pkg/command/root.go:176]: You are right. I don't think we need `getInstalledPlugins` or `getAvailablePlugins`. We can remove both of them and just use `serverPlugins, standalonePlugins, err := pluginmanager.InstalledPlugins()` to remove duplicated code.

File: packages/standalone-plugins/bundle/config/overlay/management-cluster.yaml
Comment (anujc25) [reply packages/standalone-plugins/bundle/config/overlay/management-cluster.yaml:29]: Good catch. We do not require the Context-Scoped plugins to specify `target` field. But it is absolutely important to update `packages/cliplugins/bundle/config/cliplugin-crd.yaml` with new definition. I will make the change.

**Review by marckhouzam — COMMENTED** [review]
One thing I have trouble understanding: if I write my own standalone plugin, where do I specify the target it applies to?  Should I set the "Target" field in the PluginDescriptor?

File: apis/cli/v1alpha1/cliplugin_types.go
Comment (anujc25) [reply apis/cli/v1alpha1/cliplugin_types.go:18]: Added another constant `TargetNone` and used that constant in appropriate places.

**Comment by marckhouzam** [comment]
> Currently, we do not provide `Target` information as part of `PluginDescriptor`. 

Ah ok.  We don't provide it, but the PR does add a new `Target` field to the plugin descriptor.  But I understand that it is used to pass data around, not to be filled by the plugin itself.

> We do provide that information as part of discovery using `CLIPlugin` CR for the plugin.
> 
> The reason is, `Target` information is needed by Core CLI even before we install the plugin during discovery phase. Hence, putting the `Target` information in the `PluginDescriptor` is not sufficient because `PluginDescriptor` is only available to the Core CLI once the plugin binary is downloaded.

Aha, yes that makes sense (I'm still wrapping my brain around the subtle details of this change).

So, if I create a standalone plugin that targets TMC, and I list and eventually install it using the `--local` flag, how would I specify the Target?  If this is not supported by the PR, I think it is fine, but maybe we'll need to track it for a future fix?

**Comment by chandrareddyp** [comment]
LGTM!

**Review by chandrareddyp — APPROVED** [review]
LGTM!

**Comment by marckhouzam** [comment]
@anujc25 I think we should hold off on adding the top commit "Update helptext template to update group name with plugins".

That commit is unrelated to this PR but more importantly it makes a user-visible change that I think requires a bit of discussion.  I personally would open an issue about it and post it as an independent PR so that it can get the proper attention and discussion.

**Comment by marckhouzam** [comment]
Deleting a plugin that has a target has some issues.  
As a user:
```
tanzu plugin delete management-cluster
✖  could not get plugin path for plugin "management-cluster"
```
It works if I specify the `-target` flag.  We should probably print the same message as for other commands when the user does not specify the target.

**Comment by anujc25** [comment]
>I noticed that tanzu context create ... seems to trigger plugins to be installed for all active contexts, meaning the new one I created, but also for the other target. I wonder if this is right?

This is the existing behavior with context-scoped and standalone plugins as well. I agree that it is not that great UX. As this is not exactly related to this change, I filled an issue (#4024) to track the work as a separate issue.

**Comment by anujc25** [comment]
> Deleting a plugin that has a target has some issues.

Thanks for testing this scenario. I added a commit to fix this issue.

File: cli/core/pkg/command/discovery_source.go
Comment (anujc25) [reply cli/core/pkg/command/discovery_source.go:79]: Agreed. I have filled #4025 to track this work across all available commands as this might apply to few other list commands as well.

File: packages/standalone-plugins/bundle/config/overlay/management-cluster.yaml
Comment (anujc25) [reply packages/standalone-plugins/bundle/config/overlay/management-cluster.yaml]: Added a new commit (`Update CLIPlugin CRD to 'cliplugins' and 'core-management-plugins' package`) to address this.

File: cli/core/pkg/command/root.go
Comment (marckhouzam) [inline cli/core/pkg/command/root.go:156]: This function checks for duplicate aliases between commands/plugins that are at the root level.
Now that we install plugins under the TMC target, we may want to also check for duplicate aliases at that level.
To be more future-proof, maybe it would even be better to check for duplicate aliases for each target, even though it is redundant for the K8s target (since those plugins are also at the root level, where we already check).

Comment (marckhouzam) [reply cli/core/pkg/command/root.go:156]: This can be done in a follow-up PR and does not need to delay the merge of this PR

**Comment by marckhouzam** [comment]
The last thing I wanted to check was about deleting a plugin source when we have some discoveries within a context.
It is not behaving nicely, but maybe I'm not testing properly either.
```
$ tanzu plugin source list
  NAME                       TYPE   SCOPE
  default                    local  Standalone
  default-local              local  Standalone
  default-kind-test-cluster  local  Context
$ tanzu plugin source delete default-kind-test-cluster
✖  discovery source "default-kind-test-cluster" does not exist
```
I think this is probably not even going to be used at the moment, so it does not need to block the PR.

**Review by marckhouzam — APPROVED** [review]
Excellent work @anujc25 !  This is a very nice improvement!

The PR needs to be rebased, but besides that it LGTM.

**Comment by alfredthenarwhal** [comment]
> @anujc25:  /test install-vc7
Commit: d738e7e

Tests failed! Build no: 3386

**Comment by alfredthenarwhal** [comment]
> @anujc25:  /test install-vc7
Commit: de62436

Build failed! Build no: 3388

**Review by prkalle — COMMENTED** [review]
Overall changes LGTM. Thanks!

**Comment by alfredthenarwhal** [comment]
> @anujc25:  /test install-vc7
Commit: 1c093ec

Build failed! Build no: 3393

**Review by vuil — COMMENTED** [review]
Nice set of changes!
Just a couple of minor comments on the docs side, feel free to add in the followup you have if appropriate.

Also in the testing done, it might be a good idea, after the CLI is set up to discover plugins from multiple of multiple targets, and invoking a command from a plugin with no target still works (fallback to k8s target). This is unlike the `tanzu plugin ...` plugin lcm commands, that is stricter about handling name ambiguity

**Comment by anujc25** [comment]
> Also in the testing done, it might be a good idea, after the CLI is set up to discover plugins from multiple of multiple targets, and invoking a command from a plugin with no target still works (fallback to k8s target). This is unlike the `tanzu plugin ...` plugin lcm commands, that is stricter about handling name ambiguity

Added those test details under the PR description.

File: cli/core/pkg/pluginmanager/manager.go
Comment (vuil) [inline cli/core/pkg/pluginmanager/manager.go:245]: non-blocking: It's for a corner case, but we should consider adding a TODO comment to deal with this situation:

if there are multiple discovered (but not installed) plugins of the same name then we will always end up keeping one deterministically (due to the sequence of sources/plugins that we process), but the choice is entirely arbitrary.

**Comment by alfredthenarwhal** [comment]
> @anujc25:  /test install-vc7
Commit: f466a62

Build failed! Build no: 3397

**Comment by alfredthenarwhal** [comment]
> @anujc25:  /test install-vc7
Commit: 4cabafc

Tests failed! Build no: 3403

**Comment by alfredthenarwhal** [comment]
> @anujc25:  /test install-vc7
Commit: 5221503

Tests canceled! Build no: 3408

**Comment by marckhouzam** [comment]
@anujc25 I think `tanzu config init` is broken.  I believe it has a double request to locking the config file.

**Comment by alfredthenarwhal** [comment]
> @anujc25:  /test install-vc7
Commit: fea415e

Tests passed! Build no: 3411

**Review by vuil — APPROVED** [review]
lgtm, thanks!

**Pull request merged**
